### PR TITLE
add dp/sharding gpt

### PR DESCRIPTION
**Pull request by haohongxiang**
add dp/sharding gpt

### CODE REVIEW

File: fleetx/models/gpt_model/modeling.py [hunk 1]
```diff
@@ -22,6 +22,7 @@
 from paddle.fluid import layers
 from paddle.nn.layer.transformer import _convert_param_attr_to_list
 import paddle.incubate as incubate
+from paddle.distributed.fleet.utils import recompute
 
 # import sys
```
Comment (gongweibao) [inline fleetx/models/gpt_model/modeling.py:27]: cleanup?

File: examples/gpt/group_shard/run_pretrain.py [hunk 2]
```diff
@@ -0,0 +1,235 @@
...
+            # many times. and start a new random dataloader.
+            valid_data_loader = valid_data_loader()
+            test_data_loader = test_data_loader()
+
+            # time count
+            train_reader_cost = 0.0
+            train_run_cost = 0.0
+            reader_start = time.time()
+
+            def amp_fw_bw(args, model, criterion, scaler, tokens, ids, labels,
```
Comment (gongweibao) [inline examples/gpt/group_shard/run_pretrain.py]: Move outside?

File: examples/gpt/group_shard/run_pretrain.py [hunk 3]
```diff
@@ -0,0 +1,235 @@
...
+                    loss_mbs = loss_mbs / accumulate_steps
+                    if args.use_pure_fp16:
+                        scaler.scale(loss_mbs).backward()
+                    else:
+                        loss_mbs.backward()
+                    loss = loss + loss_mbs
+                return loss
+
+            for step, batch in enumerate(train_data_loader()):
+                train_reader_cost += time.time() - reader_start
```
Comment (gongweibao) [inline examples/gpt/group_shard/run_pretrain.py:143]: It's too long to maintain.

**Review by gongweibao — CHANGES_REQUESTED** [review]

File: examples/gpt/group_shard/run_pretrain.py (hunk 2 above)
Comment (haohongxiang) [reply examples/gpt/group_shard/run_pretrain.py]: done

File: examples/gpt/group_shard/run_pretrain.py (hunk 3 above)
Comment (haohongxiang) [reply examples/gpt/group_shard/run_pretrain.py:143]: already moved to ``single/run_pretrain.py``. This function can be reused in ``data_parallel/run_pretrain.py`` and ``group_shard/run_pretrain.py``.

File: fleetx/models/gpt_model/modeling.py (hunk 1 above)
Comment (haohongxiang) [reply fleetx/models/gpt_model/modeling.py:27]: ``recompute`` is used in ``modeling.py``

**Review by gongweibao — APPROVED** [review]
LGTM

**Pull request merged**